### PR TITLE
[for 2.5.1] add mecanism to inject easily javascript code in existing tracim

### PR DESCRIPTION
**Pull request by inkhey**
related to #2524 

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

**Comment by inkhey** [comment]
> You named the feature "custom_app" in the backend code.
> If the frontend, on my branch, I have named it "instanceSpecificCode".
> Theses are too different, we need to harmonize it.
> I don't like "custom_app" because it is too generic, it doesn't differentiate this "app" from tracim's app.
> I used the name "code" because it doesn't have to be an app.
> 
> I think we should call this feature "instanceSpecificScript". I'm open to suggestions.
> 
> Note: I will do possible required frontend changes
> 
> Note2: Someone still needs to do a code quality review.

If we do instance_specific_script, what should be the path for file /instance_specific_script/ is ok for you ? Current code allow to make available all content in (renamed here) `FRONTEND__INSTANCE_SPECIFIC_SCRIPT_FOLDER_PATH` parameter in  /instance_specific_script/myfile but only .js are added to index.html as script.
-

File: backend/tracim_backend/views/frontend.py [hunk 1]
```diff
@@ -17,15 +18,29 @@
 
 
 class FrontendController(Controller):
-    def __init__(self, dist_folder_path: str):
+    def __init__(self, dist_folder_path: str, custom_toolbox_folder_path: typing.Optional[str]):
```
Comment (buxx) [inline backend/tracim_backend/views/frontend.py]:
```suggestion
    def __init__(self, dist_folder_path: str, custom_toolbox_folder_path: typing.Optional[str]) -> None:
```

File: frontend/dist/index.mak [hunk 2]
```diff
@@ -92,6 +92,10 @@
     <script type='text/javascript' src='/app/admin_workspace_user.app.js'></script>
     <script type='text/javascript' src='/app/workspace_advanced.app.js'></script>
 
+    % for custom_toolbox_file in custom_toolbox_files:
+    <script type='text/javascript' src='/custom_toolbox/${custom_toolbox_file.name}'></script>
```
Comment (Skylsmoi) [inline frontend/dist/index.mak]: add `/asset` before `/custom_toolbox`

File: backend/tracim_backend/views/frontend.py [hunk 3]
```diff
@@ -17,15 +18,29 @@
...
         self.dist_folder_path = dist_folder_path
+        self.custom_toolbox_folder_path = custom_toolbox_folder_path
+        self.custom_toolbox_files = []  # typing.List["os.DirEntry"]
```
Comment (buxx) [inline backend/tracim_backend/views/frontend.py]:
```suggestion
        self.custom_toolbox_files = []  # type: typing.List["os.DirEntry"]
```

```suggestion
        self.custom_toolbox_files = []  # type: typing.List[os.DirEntry]
```

File: backend/tracim_backend/views/frontend.py [hunk 4]
```diff
@@ -17,15 +18,29 @@
...
+            )
 
-    def _get_index_file_path(self):
+    def _get_index_file_path(self) -> str:
         index_file_path = os.path.join(self.dist_folder_path, INDEX_PAGE_NAME)
         if not os.path.exists(index_file_path):
             raise FileNotFoundError()
         return index_file_path
 
+    def _get_custom_toolboxes_files(self, custom_toolbox_dir: str) -> typing.List["os.DirEntry"]:
```
Comment (buxx) [inline backend/tracim_backend/views/frontend.py]:
```suggestion
    def _get_custom_toolboxes_files(self, custom_toolbox_dir: str) -> typing.List[os.DirEntry]:
```

Comment (inkhey) [reply backend/tracim_backend/views/frontend.py]: Not possible here. os.DirEntry is not importable in python 3.5. So if we do this, we have to import os.DirEntry and code will failed in 3.5.
see : https://stackoverflow.com/questions/38307995/create-os-direntry

File: backend/tracim_backend/views/frontend.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/views/frontend.py]: same answers as latest : this fix doesn't work on python3.5.

File: frontend/dist/index.mak (hunk 2 above)
Comment (inkhey) [reply frontend/dist/index.mak]: done. but dir is /assets/ not /asset/

**Pull request merged**